include/maliput/routing/routing_constraints.h

```diff
@@ -0,0 +1,55 @@
+// BSD 3-Clause License
+//
+// Copyright (c) 2023, Woven by Toyota. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <optional>
+#include <string>
+#include <unordered_map>
+
+namespace maliput {
+namespace routing {
+
+/// Holds the constraints that can be applied to the Router when computing a
+/// Route.
+struct RoutingConstraints {
+  /// Allows maliput::api::Lane switchs within a RoutePhase when computing the
+  /// Route.
+  bool allow_lane_switch{true};
```

Comment:
I see an omission in the API that @stonier and I struggled with in the past. The RoutePhase does not reflect which lanes the vehicle MUST occupy to meet the next transition and whether that lane is the same as the entry lane. Take, for example, a 4-lane highway where the road splits at a junction. I've attached a VERY crude drawing.  If an agent enters from the south-west and wants to travel north (taking the "left" split assuming the highway goes left to right) either lane is acceptable for the agent. Lane by lane routing will assign a lane to be preferred and the agent will dive towards the preferred lane (at least a dummy agent would). Similarly, all the lanes in the range would be listed as alternate paths, even though two of them imply a failure to route further because the agent will take the wrong fork if it is in either of the two bottom lanes. Maybe we don't care about this yet, but we may want some notion of which of the alternate LaneSRanges actually connect to the next junction on the route to give the agent better information on which lanes are actually okay.

Also, we allow lane changes, but how do we represent them semantically? This was a challenge as well. Lane-by-lane routing might choose to artificially sub-divide a RoutePhase s.t. each split sub-phase has the new target lane set as the primary LaneSRange. In this case, all agents along the route may choose to as-quickly-as-possible dive towards the new preferred lane. Not splitting, but specifying only the final lane may cause them to dive upon entering the highway.

There's no strict right answer, but we should probably augment the information at-least with which lanes will actually continue the route in each segment.

One more comment I have to make here since my wifi just died. We should keep a notion of progress along a RoutePhase so the agent can estimate how far they are from a decision point. For example, I may choose to take an alternate lane as long as I know i've got 1km to get back into this lane until my next step. That kind of sugar method will make route following easier.



![image](https://user-images.githubusercontent.com/43861892/234654228-fafad90c-8808-4ed9-840c-8f0d205e6d8a.png)

Reply:
This is a good point. We discussed it and saw what other routing APIs offer. Here are some comments to your comment to share my research and initial study of the situation.

> The RoutePhase does not reflect which lanes the vehicle MUST occupy to meet the next transition and whether that lane is the same as the entry lane.

This is good evidence that the API does not reflect my intention. I was not going to provide that as implementation. In my head, the Route will contain RoutePhases which would allow an agent to go from A to B. That imposes a constraint on the LaneSRanges that each RoutePhase can have. Let me present your example again:

![Routing 1](https://user-images.githubusercontent.com/3825465/235220992-5c834c39-d7d0-4b86-bc33-ced70cf9bea5.png)

The blue arrow represents the agent, the red square represents the target destination, each line is a Lane and each circle is a BranchPoint. The traffic flow is from West to East.

What I want to get as an answer in the Route is the following sequence of RoutePhases:

![Routing 2](https://user-images.githubusercontent.com/3825465/235221333-6a3064a6-2702-4cab-86e6-6d6dc1016551.png)

I also see how this could work too:

![Routing 3](https://user-images.githubusercontent.com/3825465/235221476-0e472849-7205-4c95-a53c-170e981cb725.png)

But in this case we are constraining the maneuver beyond what the agent can actually do and it may end up generating congestion later on. Still, I think it is a valid response.

The following constrains further the ability of the agent to arrive to the destiny. It reduces the freedom of the trajectory potentially introducing more congestion in the traffic flow.

![Routing 4](https://user-images.githubusercontent.com/3825465/235221822-eab2bcc9-12c0-4a54-bbec-87137db92184.png)

I don't expect to have any of the following responses:

![Routing 5](https://user-images.githubusercontent.com/3825465/235222160-35088be1-1bdf-45a3-9b66-470c7dd12d37.png)

![Routing 6](https://user-images.githubusercontent.com/3825465/235222170-50fa5a86-55c6-4a31-8824-54852c80a365.png)


> Also, we allow lane changes, but how do we represent them semantically? This was a challenge as well. Lane-by-lane routing might choose to artificially sub-divide a RoutePhase s.t. each split sub-phase has the new target lane set as the primary LaneSRange. In this case, all agents along the route may choose to as-quickly-as-possible dive towards the new preferred lane. Not splitting, but specifying only the final lane may cause them to dive upon entering the highway.

I think there is a bit of the answer here. I don't know a good semantic paradigm that does not constrain behavior development. This is part of the API decision actually. I think that by providing a specifc set of RoutePhases with LaneSRanges in them is also an API decision that constrains the behavior development. The agent needs to compute trajectories based on rules, traffic, kinematic and dynamic constraints, response of other agents, etc etc. I think a good choice is to provide the information and let the sophistication of the behavior implementation to decide what to consume.

> There's no strict right answer, but we should probably augment the information at-least with which lanes will actually continue the route in each segment.

Are the above pictures enough to cover the cases that you found? Do you think the expected response is good enough? What about the others?

> One more comment I have to make here since my wifi just died. We should keep a notion of progress along a RoutePhase so the agent can estimate how far they are from a decision point. For example, I may choose to take an alternate lane as long as I know i've got 1km to get back into this lane until my next step. That kind of sugar method will make route following easier.

Super interesting point. I wonder though if this should be part of the Route or be a different entity? Like a cursor or pointer that advances as the vehicle does.

Also, a preliminary question, what would it be or how would we define a decision point? I ask it because I am not sure the same definition yields for different vehicles or agents the same point in a Route or in a simulation with different sorrounding agents in the same Route.

Thanks for these comments @andrewbest-tri

Reply:
Just getting back (as I saw the ping). Yes your pictures above do answer some of my concerns, assuming that they are directed graphs. There is a piece not captured though, your graph implies that there are branch points within a straight contiguous segment of road. This is often not the case, especially in roads build in RoadRunner. You may have a long straight segment without artificial branch points in the middle in which an agent must change lanes to proceed.  I can't draw right now, but simply imagine your much nicer graph without branch points in the middle of contiguous straight lanes to see my point.

Reply:
Excellent graphs, thanks @agalbachicar. They really help drive the discussion.

> There is a piece not captured though, your graph implies that there are branch points within a straight contiguous segment of road.....but simply imagine your much nicer graph without branch points in the middle of contiguous straight lanes to see my point.

This is a good example. What is the expected response for this situation? If I were to draw graphs like you have above but without the middle branch points, I don't see how you can avoid a wrong response.

Reply:
Just for reference, the final proposal (in code - see the comments at the top of route.h) handles the situation by capturing more or less of lanes, lane start and lane end position.

* If it's got a lane and lane start, but no lane end, then the agent knows it needs to be off the lane by lane end.